tests/pytests/functional/cache/test_cache.py

```diff
@@ -0,0 +1,524 @@
+    )
+    with container.started() as factory:
+        yield factory
+
+
+@pytest.fixture(scope="module")
+def mysql_container(salt_factories, docker_client, mysql_port):
+    # TODO: Gareth has some container stuff already for lots of mysql versions - see pytests/integration/modules/test_mysql.py -W. Werner, 2021-08-05
+    container = salt_factories.get_container(
+        random_string("mysql-server-"),
+        image_name="mysql",
+        docker_client=docker_client,
+        check_ports=[mysql_port],
+        container_run_kwargs={
+            "environment": {"MYSQL_ALLOW_EMPTY_PASSWORD": "yes"},
+            "ports": {"3306/tcp": mysql_port},
+        },
+    )
+    with container.started() as factory:
+        yield factory
+
+
+@pytest.fixture(scope="module")
+def consul_container(salt_factories, docker_client, consul_port):
+    container = salt_factories.get_container(
+        random_string("consul-server-"),
+        image_name="consul",
+        docker_client=docker_client,
+        check_ports=[consul_port],
+        container_run_kwargs={"ports": {"8500/tcp": consul_port}},
+    )
+    with container.started() as factory:
+        # TODO: May want to do the same thing for redis to ensure that service is up & running
+        # TODO: THIS IS HORRIBLE. THERE ARE BETTER WAYS TO DETECT SERVICE IS UP -W. Werner, 2021-10-12
+
+        timer = Timer(timeout=10)
+        sleeptime = 0.1
+        while not timer.expired:
+            try:
+                with socket.create_connection(
+                    ("localhost", consul_port), timeout=1
+                ) as cli:
+                    cli.send(b"GET /v1/kv/fnord HTTP/1.1\n\n")
+                    cli.recv(2048)
+                    break
+            except ConnectionResetError as e:
+                if e.errno == 104:
+                    time.sleep(sleeptime)
+                    sleeptime += sleeptime
+        else:
+            assert False, "Timer expired before connecting to consul"
+        yield factory
+
+
+@pytest.fixture
+def redis_cache(minion_opts, redis_port, redis_container):
+    opts = minion_opts.copy()
+    opts["cache"] = "redis"
+    opts["cache.redis.host"] = "127.0.0.1"
+    opts["cache.redis.port"] = redis_port
+    # NOTE: If you would like to ensure that alternate prefixes are properly
+    # tested, simply change these values and re-run the tests.
+    opts["cache.redis.bank_prefix"] = "#BANKY_BANK"
+    opts["cache.redis.bank_keys_prefix"] = "#WHO_HAS_MY_KEYS"
+    opts["cache.redis.key_prefix"] = "#LPL"
+    opts["cache.redis.timestamp_prefix"] = "%TICK_TOCK"
+    opts["cache.redis.separator"] = "\N{SNAKE}"
+    cache = salt.cache.factory(opts)
+    yield cache
+
+
+@pytest.fixture(scope="module", autouse="true")
+def ensure_deps(states):
+    installation_result = states.pip.installed(
+        name="fnord",
+        pkgs=["python-etcd", "redis", "redis-py-cluster", "python-consul", "pymysql"],
```

Comment:
While I see what you're trying to do here, keep requirements in a single place, this won't pin to specific CI versions, which, from time to time, might break.
Why aren't we adding these requirements to the CI requirements files?

Reply:
They could (should?) probably go there as well.

Actually what I was going for is to mandate that this test runs, instead of getting skipped because requirements aren't installed.

Though, now that I'm thinking about it... for these modules I think it's actually a better approach to ensure that whatever is the latest dependency is tested, so that if (when) something breaks we get advanced notice of it. The alternative is for the documentation for these modules to specify that they've only been tested with a particular version of the dependencies.